fairscale/optim/oss.py

```diff
@@ -91,6 +104,98 @@ def state_dict(self) -> dict:
+        local_state = self.state_dict()
+
+        if len(local_state["state"]) == 0:
+            return []
+
+        for rank in range(dist.get_world_size(group=self.group)):
+            if rank == self.rank:
+                logging.info("Saving self state")
+                global_optim_state.append(
+                    recursive_copy_to_device(
+                        local_state, non_blocking=True, device=torch.device("cpu")
+                    )
+                )
+
+                # Sync with other replicas
+                broadcast_object(empty_buffer, src_rank=rank)
+            else:
+                # Reuse the param_groups from this rank, these are shared across replicas
+                logging.info("Receiving state from rank %s ", rank)
+                replica_state = {
+                    "state": broadcast_object(empty_buffer, src_rank=rank),
+                    "param_groups": local_state["param_groups"],
+                }
+
+                # Fetch from the other replicas
+                global_optim_state.append(
+                    recursive_copy_to_device(
+                        replica_state, non_blocking=True, device=torch.device("cpu")
+                    )
+                )
+                logging.info("State from rank %s received", rank)
+
+        return global_optim_state
+
+    def _broadcast_state_dict(self) -> None:
+        """
+        Broadcast this rank's state shard, discard others
+        """
+        empty_buffer = torch.empty([1], dtype=torch.uint8)
+        local_state = self.state_dict()
+
+        if len(local_state["state"]) == 0:
+            return
+
+        for rank in range(dist.get_world_size(group=self.group)):
+            if rank == self.rank:
+                # Send the state to the reference replica
+                logging.info(
+                    "Sending the sharded SGD state to the reference replica from rank %s",
+                    rank,
+                )
+                broadcast_object(local_state["state"], src_rank=rank)
+            else:
+                # Discard this tensor/rank, broadcast necessary for syncing
+                logging.info("Discarding broadcast from rank %s", rank)
+                broadcast_object(empty_buffer, src_rank=rank)
+
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> List[dict]:
+        """ Update the consolidated state_dict list, one per rank.
+
+        This needs to be called on all replicas """
+
+        if self.rank == recipient_rank:
+            # Pull the sharded state from all the other replicas
+            # Store all the states in order, rank by rank
+            logging.info("Pulling the sharded SGD state from all replicas")
+            self._global_state_dict = self._collect_state_dict()
+        else:
+            # Acknowledge broadcasts, and send this rank's shard when needed
+            self._broadcast_state_dict()
+
+    @property
+    def global_state_dict(self):
```

Comment:
Should this be called state_dict and state_dict renamed to local_state_dict? This way OSS nearly just works as a drop-in replacement.